target/linux/ramips/dts/mt7621_tozed_zlt-s12-pro.dts

```diff
@@ -0,0 +1,245 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+
+#include "mt7621.dtsi"
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+#include <dt-bindings/leds/common.h>
+
+/ {
+	compatible = "tozed,zlt-s12-pro", "mediatek,mt7621-soc";
+	model = "TOZED ZLT S12 PRO";
+
+	aliases {
+		led-boot = &led_power_blue;
+		led-failsafe = &led_power_yellow;
+		led-running = &led_power_blue;
+		led-upgrade = &led_power_blue;
+		label-mac-device = &wifi1;
+	};
+
+	chosen {
+		bootargs = "console=ttyS0,115200";
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		led-modem-blue {
+			label = "blue:modem";
+			gpios = <&gpio 33 GPIO_ACTIVE_LOW>;
+		};
+
+		led-modem-red {
+			label = "red:modem";
+			gpios = <&gpio 27 GPIO_ACTIVE_LOW>;
+		};
+
+		led_power_blue: led-power-blue {
+			label = "blue:power";
+			gpios = <&gpio 16 GPIO_ACTIVE_LOW>;
+		};
+
+		led_power_yellow: led-power-yellow {
+			label = "yellow:power";
+			gpios = <&gpio 29 GPIO_ACTIVE_LOW>;
+		};
+
+		led-wifi {
+			label = "blue:wifi";
+			gpios = <&gpio 28 GPIO_ACTIVE_LOW>;
+			linux,default-trigger = "phy0tpt";
+		};
+
+		led-signal1 {
+			label = "blue:signal1";
+			gpios = <&gpio 25 GPIO_ACTIVE_LOW>;
+		};
+
+		led-signal2 {
+			label = "blue:signal2";
+			gpios = <&gpio 24 GPIO_ACTIVE_LOW>;
+		};
+
+		led-signal3 {
+			label = "blue:signal3";
+			gpios = <&gpio 23 GPIO_ACTIVE_LOW>;
+		};
+
+		led-signal4 {
+			label = "blue:signal4";
+			gpios = <&gpio 22 GPIO_ACTIVE_LOW>;
+		};
+
+		led-signal5 {
+			label = "blue:signal5";
+			gpios = <&gpio 26 GPIO_ACTIVE_LOW>;
+		};
+
+		led-phone {
+			label = "blue:phone";
+			gpios = <&gpio 31 GPIO_ACTIVE_LOW>;
+		};
+	};
+
+	keys {
+		compatible = "gpio-keys";
+
+		button-reset {
+			label = "reset";
+			gpios = <&gpio 13 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+		};
+
+		button-rfkill {
+			label = "rfkill";
+			gpios = <&gpio 15 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RFKILL>;
+		};
+
+		button-wps {
+			label = "wps";
+			gpios = <&gpio 18 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_WPS_BUTTON>;
+		};
+	};
+
+	gpio_export {
+		compatible = "gpio-export";
+
+		lt72_ant0 {
+			gpio-export,name = "lt72_ant0";
+			gpio-export,input = <0>;
+			gpios = <&gpio 6 GPIO_ACTIVE_HIGH>;
+		};
+
+		lt72_ant1 {
+			gpio-export,name = "lt72_ant1";
+			gpio-export,input = <0>;
+			gpios = <&gpio 7 GPIO_ACTIVE_HIGH>;
+		};
+
+		lt72_power {
+			gpio-export,name = "lt72_power";
+			gpio-export,output = <1>;
+			gpios = <&gpio 14 GPIO_ACTIVE_HIGH>;
+		};
+	};
+};
+
+&spi0 {
+	status = "okay";
+
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <10000000>;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "u-boot";
+				reg = <0x0 0x30000>;
+				read-only;
+			};
+
+			partition@30000 {
+				label = "u-boot-env";
```

Comment:
Is it possible to access this u-boot environment from Linux?

Reply:
This is doable with using the uboot-envtools package. The variables won't appear properly on this specific device, so I'm not going to include that.

I also think the `NVMEM_U_BOOT_ENV` driver is the way to go forward which lets you define the uboot-env partition on the devicetree instead of the OpenWrt-specific `package/boot/uboot-envtools/files/`. The fw_setenv and fw_printenv commands from the uboot-envtools package should start using the driver in the future. The driver hasn't got writing ability (@reg_write) as of now so that needs to be addressed first. @rmilecki let me know your thoughts.

Reply:
I see it is already implemented. https://github.com/u-boot/u-boot/commit/8142c4554ffaa927529f24427a35f7ee2861793a

We should make the uboot-envtools package on OpenWrt follow suit.

Reply:
The patch should be included in the u-boot envtools version we are using in master. I have never tried how it works. Maybe we have to adapt some scripts to make it fully work in OpenWrt.